frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -512,36 +518,80 @@ def _call_with_quantum_ctx(self, ctx):
         in_classical_tracers = self.preds
         regions: List[HybridOpRegion] = []
 
-        out_treedefs, out_signatures = [], []
-        for branch in self.branch_fns + [self.otherwise_fn]:
+        in_sigs, out_sigs = [], []
+        for branch_fn in self.branch_fns + [self.otherwise_fn]:
             quantum_tape = QuantumTape()
+            wfun, in_sig, out_sig = deduce_signatures(
+                branch_fn, [], {}, expansion_strategy=self.expansion_strategy
+            )
+            assert len(in_sig.in_type) == 0
             with EvaluationContext.frame_tracing_context(ctx) as inner_trace:
-                wffa, _, _, out_tree = deduce_avals(branch, [], {})
                 with QueuingManager.stop_recording(), quantum_tape:
-                    res_classical_tracers = [inner_trace.full_raise(t) for t in wffa.call_wrapped()]
-                    res_avals = [shaped_abstractify(res) for res in res_classical_tracers]
-            regions.append(HybridOpRegion(inner_trace, quantum_tape, [], res_classical_tracers))
-            out_treedefs.append(out_tree())
-            out_signatures.append(res_avals)
-
-        _assert_cond_result_structure(out_treedefs)
-        _assert_cond_result_types(out_signatures)
-        out_classical_tracers = [new_inner_tracer(outer_trace, aval) for aval in out_signatures[0]]
-        self._operation = Cond(in_classical_tracers, out_classical_tracers, regions)
-        return tree_unflatten(out_tree(), out_classical_tracers)
-
-    def _call_with_classical_ctx(self):
-        args, args_tree = tree_flatten([])
-        args_avals = tuple(map(_abstractify, args))
-        branch_jaxprs, consts, out_treedefs = initial_style_jaxprs_with_common_consts1(
-            (*self.branch_fns, self.otherwise_fn), args_tree, args_avals, "cond"
+                    res_classical_tracers = [inner_trace.full_raise(t) for t in wfun.call_wrapped()]
+
+            regions.append(
+                HybridOpRegion(
+                    inner_trace,
+                    quantum_tape,
+                    [],
+                    collapse(out_sig.out_type(), res_classical_tracers),
+                )
+            )
+            in_sigs.append(in_sig)
+            out_sigs.append(out_sig)
+
+        _assert_cond_result_structure([s.out_tree() for s in out_sigs])
+        _assert_cond_result_types([[t[0] for t in s.out_type()] for s in out_sigs])
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # TODO: Creating output tracers like this is not quite cocrect. We need to calculate the
+        # unified result type first and only then use it as the output type of this primitive.
```

Comment:
I have improved the comment, please check. This is a rather complex and unfortunate problem which appeared because we do result unification across the conditional branches. Typically, we need to re-trace the whole primitive after applying user-defined transformations that involve changing types, but our code does not support this kind of processing currently.
So we take the last branch that becomes our guess for what type is going to survive.

Reply:
I'm sorry, I lack a bit of context here. What does "result unification" mean? Does that just mean type checking the result types of the return branches?

Reply:
No it is not only about checking. We apply Jax-based type conversion. I have updated the comment with the code example, see the [new version of the comment](https://github.com/PennyLaneAI/catalyst/blob/427d5ad7d623d7913954cbd1df3614d1ca735a71/frontend/catalyst/api_extensions/control_flow.py#L540)